src/nodes/Alias.ts

```diff
@@ -38,21 +38,38 @@ export class Alias extends NodeBase {
    * Resolve the value of this alias within `doc`, finding the last
    * instance of the `source` anchor before this node.
    */
-  resolve(doc: Document): Scalar | YAMLMap | YAMLSeq | undefined {
+  resolve(
+    doc: Document,
+    ctx?: ToJSContext
+  ): Scalar | YAMLMap | YAMLSeq | undefined {
     let found: Scalar | YAMLMap | YAMLSeq | undefined = undefined
-    visit(doc, {
-      Node: (_key: unknown, node: Node) => {
-        if (node === this) return visit.BREAK
-        if (node.anchor === this.source) found = node
-      }
-    })
+
+    if (!ctx) {
+      visit(doc, {
+        Node: (_key: unknown, node: Node) => {
+          if (node === this) return visit.BREAK
+          if (node.anchor === this.source) found = node
+        }
+      })
+      return found
+    }
+
+    if (!ctx.aliasResolveCache) {
+      ctx.aliasResolveCache = buildCache(doc)
+    }
+
+    for (const node of ctx.aliasResolveCache) {
+      if (node === this) break
+      if (node.anchor === this.source) found = node
+    }
+
     return found
   }
```

Comment:
We can presume that calling `resolve()` without a context is really rare, so this can be simpler, something like:
```ts
  resolve(
    doc: Document,
    ctx?: ToJSContext
  ): Scalar | YAMLMap | YAMLSeq | undefined {
    let nodes: Node[]
    if (ctx?.aliasResolveCache) {
      nodes = ctx.aliasResolveCache
    } else {
      nodes = []
      visit(doc, {
        Node: (_key: unknown, node: Node) => {
          if (isAlias(node) || node.anchor) nodes.push(node)
        }
      })
      if (ctx) ctx.aliasResolveCache = nodes
    }

    let found: Scalar | YAMLMap | YAMLSeq | undefined = undefined
    for (const node of nodes) {
      if (node === this) break
      if (node.anchor === this.source) found = node
    }
    return found
  }
```

Reply:
Sure, looks a lot better. Done